api.ts

```diff
@@ -1271,8 +1282,50 @@ export type MetadataFormula = BaseFormula<[ParamDef<Type.string>, ParamDef<Type.
   schema?: any;
 };
 
+type GenericMetadataFormula = BaseFormula<ParamDefs, any> & {
+  schema?: any;
+};
+
+/**
+ * These will be created with a helper function.
+ */
+interface PropertyAutocompleteFormattedResult {
```

Comment:
Is this helper not implemented yet? I'm missing where the object of this shape is constructed from the results that the pack maker returns. I guess I would have expected it to happen in the thunk.

Reply:
Yeah, I haven't implement a helper yet, just improvising this type in the PR. Not sure exactly how things should look on the pack side, from an api-perspective.

Reply:
I don't think the pack should be able to return `any`. Seems like it should follow the example of regular autocomplete, where you can either return a scalar, or you can return a `{display, value}` object if you want to do something fancy. Wouldn't we want them to explicit return a `display` string when returning an object, too?

If so, I'd suggest that `PropertyAutocompleteMetadataFunction` returns `string | number | boolean | PropertyAutocompleteFormattedResult`.

Reply:
Parameter autocomplete only supports `string | number` as the value, while property autocomplete here is aspiring to support at least three scenarios:

1) Scalars like `string`, `number` and potentially Dates or booleans (like parameter autocomplete)
2) Scalars with user-friendly names like `{display: 'John', value: 'johnsUserId1234'}` (like parameter autocomplete)
3) Objects like `{someProperty: 'someValue', 'anotherProperty': 123}` (unlike parameter autocomplete)

The challenge is distinguishing between (2) and (3). One way to handle that could be like this:

```
interface SimplePropertyAutocompleteResults {
  type: 'Simple',
  cacheTtlSecs?: number;
  results: Array<any>;
}
interface AnnotatedPropertyAutocompleteResults {
  type: 'Annotated',
  cacheTtlSecs?: number;
  results: Array<PropertyAutocompleteFormattedResult>;
}
type PropertyAutocompleteResults = SimplePropertyAutocompleteResults | AnnotatedPropertyAutocompleteResults;
```

A tradeoff there would be that with this approach we'd probably want people to use a helper function most of the time in the pack, something like this:

```
  return coda.makeSimplePropertyAutocompleteResults({catchTtlSecs: 3}, ['foo', 'bar']);

  // With display values
  return coda.makeAnnotatedPropertyAutocompleteResults([
    {display: 'foo', value: 'fooId123'},
    {display: 'bar', value: 'barId456'},
  ]);
```

That's probably fine?... cc @ekoleda-codaio for your thoughts

I was originally trying to make it so that in the simplest case you could just `return ['foo', 'bar']` without calling a helper function, which would mean more runtime guessing about whether something is or isn't a `PropertyAutocompleteFormattedResult`

Reply:
Can't we just say that for (3) you have to return `{display: 'someDisplay', value: {someProperty: 'someValue', 'anotherProperty': 123}}` and take the guesswork out of it?

It seems like we probably need an explicit display value when the pack is returning an object value result anyway, for various render scenarios, in which case we have to use this result format anyway?

Reply:
You don't necessarily need an explicit separate display value for returning an object. For a concrete example, here a pack prototype code for autocompleting Jira Users (and Statuses) in `formulas.ts`: https://github.com/coda/packs/compare/dweitzman-jira-autocomplete

The returned autocomplete result for `assignee` (a `User` property) has this type `SchemaType<typeof User>[]`, where `User` is the `makeObjectSchema` for a user, so it's returning a result like `[{displayName: 'John', accountId: 'abc123'}]`

The prototype UI doesn't support chip rendering yet but knows what text to render (`John`) since `User` has `displayProperty: 'displayName'`

Reply:
Ah ok. So it sounds like the legal returns values are scalar OR {display, value} OR object that matches the object schema of this property in the row schema?

What are the consequences if we see {display, value} and interpret it as an annotated display value rather than an object? If it's not consequential then maybe it doesn't matter if we just guess. If we know the schema we're looking for, maybe it's even easier to guess well.

Otherwise I don't think it's unreasonable to have to wrap one of the object types to disambiguate.